tamr_unify_client/dataset/collection.py

```diff
@@ -171,6 +171,189 @@ def create_from_dataframe(
 
         return dataset
 
+    def create_from_geodataframe(
+        self, geodf, primary_key_name, dataset_name, geo_attr="geometry"
+    ):
+        """Creates a dataset in this collection with the given name, creates an attribute for each column in the `geodf`
+        (with `primary_key_name` as the key attribute), and upserts a record for each row of `geodf`.
+
+        Each attribute has the default type `ARRAY[STRING]`, besides the key attribute, which will have type `STRING`,
+        and the geometry attribute `geo_attr` which will have the type `GEOSPATIAL`.
+
+        This function attempts to ensure atomicity, but it is not guaranteed. If an error occurs while creating
+        attributes or records, an attempt will be made to delete the dataset that was created. However, if this
+        request errors, it will not try again.
+
+        :param geodf: The data to create the dataset with.
+        :type df: :class:`geopandas.geodataframe.GeoDataFrame`
+        :param primary_key_name: The name of the primary key of the dataset. Must be a column of `geodf`.
+        :type primary_key_name: str
+        :param dataset_name: What to name the dataset in Tamr. There cannot already be a dataset with this name.
+        :type dataset_name: str
+        :param geo_attr: The name of the geometry column within the `geodf`. Must be a column of `geodf`. Defaults
+            to `geometry`.
+        :type geo_attr: str
+        :returns: The newly created dataset.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.Dataset`
+        :raises KeyError: If `primary_key_name` is not a column in `df`.
+        :raises CreationError: If a step in creating the dataset fails.
+        """
+        if primary_key_name not in geodf.columns:
+            raise KeyError(f"{primary_key_name} is not an attribute of the data")
+
+        creation_spec = {"name": dataset_name, "keyAttributeNames": [primary_key_name]}
+        try:
+            dataset = self.create(creation_spec)
+        except HTTPError:
+            raise CreationError("Dataset was not created")
+        # after this point, if a request fails, try to undo the change by deleting this dataset
+
+        attributes = dataset.attributes
+        for col in geodf.columns:
+            if col == primary_key_name:
+                # this attribute already exists, so don't create it again
+                continue
+
+            if col == geo_attr:
```

Comment:
Can there be more than one geo attribute in a Geo DataFrame? If so, we should account for that here. Tamr is certainly able to handle datasets with multiple geo attributes.

Also is there a way to detect Geo attributes by their `dtype` or similar in GeoPandas so we can always do the right thing instead of relying on users to pass in the correct names?

Reply:
Yes, it can have more than one geo_attribute (encoded as `geoSeries`) in one geoDF, but only one will be the 'active' in terms of which one is called when calling methods on the geoDF. 

Default will be `geometry` when reading in most file types in geopandas, but if people have multiple types, it should be set explicitly.

Reply:
I've now included a detection for geometry columns, which will be passed a spec of geometry type.

Reply:
Hmm, this is where I am out of my depth regarding geospatial stuff. @nbateshaus is limitation on having at most one geo attribute something we impose in `tamr-client` or something imposed by our underlying dependencies?

In either, case what do you recommend?

Reply:
It's possible to convert the geom that isn't the primary one to a string and upsert to tamr; not sure if that's a solution.

Reply:
@nbateshaus Any thoughts on how you would want this to work with the `tamr-client`. For now it uses `dataset.from_geo_features()` but this does not allow for upserting multiple geometries?

The way I see it we can:
[ ] Implement this now for only one geometry to be used, ignoring others.
[ ] Implement this now for only one geometry to be used, upserting the others as string
[ ] Implement multiple geometries in `dataset.from_geo_features()` and then use that here. 

The last one seems difficult as I understand it uses the `__geo_interface__`.

Reply:
`__geo_interface__` is restricted to a single geometry, so we should not use that.
We should use logic similar to `from_geo_features()` to get all the geometries into the geodataframe.

More broadly, `__geo_interface__` is a manifestation of Simple Feature Access which says that a given "feature" can have at most one geometry. That is only one specific standard for working with geospatial information.

Geodataframe (and Tamr Tamr) are less restrictive. Just as we built interoperability between Tamr dataset and `__geo_interface__`, we should build interoperability between Tamr dataset and geodataframe - but this will have to be a completely separate implementation, separate from anything `__geo_interface__` related.

Reply:
@nbateshaus 
I agree on the interoperability; this wrapper creates multiple geospatial attributes in a new `dataset` based on the type within geopandas. My issue is with upserting/serialising these. 

I run into issues with `from_geo_features()`. Although the docs mention it is supposed to work with multiple geometries, I understand it only upserts one geometry column feature per record (given the underlying call to `_feature_to_record()` which takes `geo_attr: The **singluar** attribute on the record to use for the geometry`.)

So the current implementation of `from_geo_features()` appears to ignore all other geometry attributes in the GDF. This implementation carries this forward. The question is more whether we want to upsert multiple geometries per record; which would seem to require more changes. 

There's also the very probable case that I've missed something essential, in which case an example of how to upsert multiple geometries per records would be great so I can make the changes 😊